x-pack/plugins/xpack_main/server/lib/populate_ui_capabilities.ts

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import _ from 'lodash';
+import { UICapabilities } from 'ui/capabilities';
+import { Feature } from '../../types';
+
+export function populateUICapabilities(
+  xpackMainPlugin: Record<string, any>,
+  injectedVars: Record<string, any>
+): UICapabilities {
+  const features: Feature[] = xpackMainPlugin.getFeatures();
+
+  const featureCapabilities: UICapabilities[] = features.map(getCapabilitiesFromFeature);
+
+  return mergeCapabilities(injectedVars.uiCapabilities || {}, ...featureCapabilities);
+}
+
+function getCapabilitiesFromFeature(feature: Feature): UICapabilities {
+  const capabilities: UICapabilities = {
+    navLinks: {},
+    [feature.id]: {},
+  };
+
+  if (feature.navLinkId) {
+    capabilities.navLinks[feature.navLinkId] = true;
+  }
+
+  const featureCapabilities: Record<string, boolean> = Object.values(feature.privileges).reduce(
+    (acc, privilege) => {
+      return {
+        ...acc,
+        ...privilege.ui.reduce(
+          (privilegeAcc, capabillity) => ({
+            ...privilegeAcc,
+            [capabillity]: true,
+          }),
+          {}
+        ),
+      };
+    },
+    {}
+  );
+
+  capabilities[feature.id] = featureCapabilities;
+
+  return capabilities;
+}
+
+function mergeCapabilities(...allCapabilities: UICapabilities[]): UICapabilities {
+  return allCapabilities.reduce(
+    (acc, capabilities) => {
+      const featureCapabilities = Object.keys(capabilities)
+        .filter(key => key !== 'navLinks')
```

Comment:
`filter` isn't strictly necessary, because `navLinks` are always overwritten below, but I figured it'd be slightly less work for Kibana to have to do